### PR TITLE
Add ability to set a custom access checker when using CommandSpec

### PR DESCRIPTION
**Pull request by CrystalMageX**
Instead of only allow *permissions* to be checked in `CommandSpec`, this will allow plugin developers to supply an *custom* ~~`Predicate<CommandSource>`~~ `Function<CommandSource, Optional<Text>>` instead. This is a **better idea** than putting the code in the beginning of the `CommandExecutor`, because the access control checking can be kept in a **central location** now, at `testPermission(CommandSource)`. When the returned `Optional`'s `Text` is not present, it means that there is no error message, so no error.

### CODE REVIEW

**Comment by liach** [comment]
I like this, and I suggest you to squash commits.

**Comment by kashike** [comment]
@CrystalMageX Why not call this method `testPermission` as well?

File: src/main/java/org/spongepowered/api/command/spec/CommandSpec.java [hunk 1]
```diff
@@ -113,7 +113,21 @@ private Builder() {}
...
+            return this;
+        }
+
+        /**
+         * Set the permission tester that will be checked before using this command.
+         *
+         * @param tester The permission tester to consult
+         * @return this
+         */
+        public Builder permissionTester(Function<CommandSource, Optional<Text>> tester) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/command/spec/CommandSpec.java]: This should just accept a `Predicate` in my opinion. Add a new method for the error message, will allow it to be handled nicer as well.

File: src/main/java/org/spongepowered/api/command/spec/CommandSpec.java [hunk 2]
```diff
@@ -266,9 +281,12 @@ public CommandSpec build() {
      * @throws CommandException if the source does not have permission
      */
     public void checkPermission(CommandSource source) throws CommandException {
-        checkNotNull(source, "source");
-        if (!testPermission(source)) {
-            throw new CommandPermissionException();
+        if (this.permissionTester != null) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/command/spec/CommandSpec.java]: See above comment.

**Comment by Zidane** [comment]
I don't want to pull this without a handy singleton Function somewhere that keeps the same functionality present here as many devs won't care to do more indepth permission checks.

**Comment by CrystalMageX** [comment]
@kashike Having a `Function` instead of a `Predicate` would avoid having redundant code repeating the same code.

Consider the `Predicate` version of a magic plugin's */fireball* command:
```
public boolean hasAccess(CommandSource src) {
    if (!(src instanceof Player)) return false;
    Player p = (Player) src;

    if (MagicalFactions.WIZARD != getFaction(p)) return false;
    if (getLevel(p) < 10) return false;

    return true;
}

public Optional<Text> getError(CommandSource src) {
    if (src instanceof Player)
        return Optional.of(Text.of("You must be a player to cast spells!"));
    Player p = (Player) src;

    if (MagicalFactions.WIZARD != getFaction(p))
        return Optional.of(Text.of("You must be a wizard to firewall players!"));
    if (getLevel(p) < 10)
        return Optional.of(Text.of("You aren't a wise enough wizard yet for this skill!"));

    // should never happen and only needed in this Predicate version
    throw new IllegalStateException("???");
}
```

Consider the same command with a `Function`:
```
public Optional<Text> hasAccess(CommandSource src) {
    if (!(src instanceof Player))
        return Optional.of(Text.of("You must be a player to cast spells!"));
    Player p = (Player) src;

    if (MagicalFactions.WIZARD != getFaction(p))
        return Optional.of(Text.of("You must be a wizard to firewall players!"));
    if (getLevel(p) < 10)
        return Optional.of(Text.of("You aren't a wise enough wizard yet for this skill!"));

    // No error message, not even a generic one, means no error
    return Optional.empty();
}
```

The `Function` version is **9** lines shorter, does not repeat code that could have put bugs in multiple places, and does not have to worry about sending an error message when there is no error.

@Zidane The method `CommandSpec.builder().permission("some.permission.node")` is still present. (https://github.com/SpongePowered/SpongeAPI/pull/1080/files#diff-4201653595f5b8d8b5f941fef8403bdbR115). It has only be re-written *internally* to use the `permissionTester` field, to simplify the code. All current plugins using the `CommandSpec.builder().executor(commandExecutor).permission("pluginName.category,permission").build()` syntax will **still work**. 

> I don't want to pull this without a handy singleton Function somewhere that keeps the same functionality present here as many devs won't care to do more indepth permission checks.

The "handy singleton Function", `Builder.permission()`, remains available, and developers that don't care about more in depth permission checks can safely ignore `Builder.permissionTester()` without any effect on the behavior of their plugin.

**Comment by Zidane** [comment]
@CrystalMageX 

Upon reviewing @kashike 's proposed change and your rebuttal, I still agree with @kashike solely on the premise that I seriously dislike the fact that a present Text object in the Optional means that they failed the permission check. That reeks of bad design.

Closing this as I think it needs re-thinking.

**Pull request closed without merging**